lib/pure/times.nim

```diff
@@ -1396,228 +1488,721 @@ proc `*=`*[T: TimesMutableTypes, U](a: var T, b: U) =
+    let pattern = f.patterns[patIdx].LayoutPattern
+    case pattern
+    of Lit:
+      patIdx.inc
+      let len = f.patterns[patIdx]
+      patIdx.inc
+      for _ in 1'u8..len:
+        if input[inpIdx] != f.patterns[patIdx].char:
+          raiseParseException(f, input,
+                              "Unexpected character: " & input[inpIdx])
+        inpIdx.inc
+        patIdx.inc
+    else:
+      if not parsePattern(input, pattern, inpIdx, parsed):
+        raiseParseException(f, input, &"Failed on pattern '{pattern}'")
+      patIdx.inc
+
+  if inpIdx <= input.high:
+    raiseParseException(f, input,
+                        "Parsing ended but there was still input remaining")
+
+  if patIdx <= f.patterns.high:
+    raiseParseException(f, input,
+                        "Parsing ended but there was still patterns remaining")
+
+  result = toDateTime(parsed, zone, f, input)
+
+proc parse*(input, layout: string, tz: Timezone = local()): DateTime =
+  ## Shorthand for constructing a ``TimeLayout`` and using it to parse
+  ## ``input`` as a ``DateTime``.
+  ##
+  ## See `Parsing and formatting dates`_ for documentation of the
+  ## ``layout`` argument.
+  runnableExamples:
+    let dt = initDateTime(01, mJan, 2000, 00, 00, 00, utc())
+    doAssert dt == parse("2000-01-01", "yyyy-MM-dd", utc())
+  let dtFormat = initTimeLayout(layout)
+  result = dtFormat.parse(input, tz)
+
+proc parse*(input: string, layout: static[string], zone: Timezone = local()): DateTime =
+  ## Overload that validates ``layout`` at compile time.
+  const f = initTimeLayout(layout)
+  result = f.parse(input, zone)
+
+proc parseTime*(input, layout: string, zone: Timezone): Time =
+  ## Shorthand for constructing a ``TimeLayout`` and using it to parse
+  ## ``input`` as a ``DateTime``, then converting it a ``Time``.
+  ##
+  ## See `Parsing and formatting dates`_ for documentation of the
+  ## ``layout`` argument.
+  runnableExamples:
+    let tStr = "1970-01-01T00:00:00+00:00"
+    doAssert parseTime(tStr, "yyyy-MM-dd'T'HH:mm:sszzz", utc()) == fromUnix(0)
+  parse(input, layout, zone).toTime()
+
+proc parseTime*(input: string, layout: static[string], zone: Timezone): Time =
+  ## Overload that validates ``layout`` at compile time.
+  const f = initTimeLayout(layout)
+  result = f.parse(input, zone).toTime()
+
+#
+# End of parse & format implementation
+#
 
 proc `$`*(dt: DateTime): string {.tags: [], raises: [], benign.} =
   ## Converts a `DateTime` object to a string representation.
   ## It uses the format ``yyyy-MM-dd'T'HH-mm-sszzz``.
   runnableExamples:
     let dt = initDateTime(01, mJan, 2000, 12, 00, 00, utc())
-    doAssert $dt == "2000-01-01T12:00:00+00:00"
-  try:
-    result = format(dt, "yyyy-MM-dd'T'HH:mm:sszzz") # todo: optimize this
-  except ValueError: assert false # cannot happen because format string is valid
+    doAssert $dt == "2000-01-01T12:00:00Z"
+  result = format(dt, "yyyy-MM-dd'T'HH:mm:sszzz") # todo: optimize this
```

Comment:
Does ``format`` no longer raise? Also, maybe we can just remove that "TODO" now.

Reply:
The `static[T]` overloads means that errors in the format string are cough at compile time, so if the format string is known at compile time `format` wont raise any exception